include/swift/AST/AccessScope.h

```diff
@@ -0,0 +1,89 @@
+//===--- AccessScope.h - Swift Access Scope ---------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the AccessScope class.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_ACCESSSCOPE_H
+#define SWIFT_ACCESSSCOPE_H
+
+#include "llvm/ADT/PointerIntPair.h"
+#include "swift/AST/AttrKind.h"
+#include "swift/AST/DeclContext.h"
+
+namespace swift {
+
+/// AccessScope - The wrapper around the outermost DeclContext from which
+/// a particular declaration can be accessed.
+class AccessScope {
+  /// \brief The declaration context (if not public) along with a bit saying
+  /// whether this scope is private (or not).
+  llvm::PointerIntPair<const DeclContext *, 1, bool> Value;
+public:
+  AccessScope(const DeclContext *DC = nullptr, bool isPrivate = false)
+    : Value(DC, isPrivate) {}
+
+  void setPrivate(bool isPrivate) { Value.setInt(isPrivate); }
+
+  /// /returns nullptr if access scope is public.
+  const DeclContext *getDeclContext() const { return Value.getPointer(); }
+
+  bool operator==(const AccessScope RHS) const {
```

Comment:
For better or for worse, we don't use `const` on things passed or returned by value. Please remove it in all locations.